python/ray/data/dataset.py

```diff
@@ -613,6 +613,58 @@ def do_shuffle(block_list, clear_input_blocks: bool, block_udf, remote_args):
         )
         return Dataset(plan, self._epoch, self._lazy)
 
+    def random_sample(
+        self, fraction: float, *, seed: Optional[int] = None
+    ) -> "Dataset[T]":
+        """Randomly samples a fraction of the elements of this dataset.
+
+        Note that the exact number of elements returned is not guaranteed,
+        and that the number of elements being returned is roughly fraction * total_rows.
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.range(100) # doctest: +SKIP
+            >>> ds.random_sample(0.1) # doctest: +SKIP
+            >>> ds.random_sample(0.2, seed=12345) # doctest: +SKIP
+
+        Args:
+            fraction: The fraction of elements to sample.
+            seed: Seeds the python random pRNG generator.
+
+        Returns:
+            Returns a Dataset containing the sampled elements.
+        """
+        import random
+        import pyarrow as pa
+        import pandas as pd
+
+        if self.num_blocks() == 0:
+            raise ValueError("Cannot sample from an empty dataset.")
+
+        if fraction < 0 or fraction > 1:
+            raise ValueError("Fraction must be between 0 and 1.")
+
+        if seed:
+            random.seed(seed)
+
+        def process_batch(batch):
```

Comment:
Perf nit: we could avoid unnecessary `O(len(batch))` scans/iterations and save on space by doing streaming Bernoulli downsampling for the list case and streaming mask construction for the Arrow case:
```python
def process_batch(batch):
    if isinstance(batch, list):
        return [row for row in batch if random.random() <= fraction]
    if isinstance(batch, pa.Table):
        return batch.filter(pa.array(random.random() <= fraction for _ in range(len(batch))))
    if isinstance(batch, pd.DataFrame):
        return batch.sample(frac=fraction)
    raise ValueError("Unsupported batch type: {}".format(type(batch)))
```
The Arrow case won't be streaming, but it will construct a boolean binary array that's bitpacked (low space overhead) and the filter will be vectorized, so still pretty good.

Reply:
Will take care of this in a moment. I do not have any experience with arrow since I've mostly been working on datasets that can fit on one GPU.

Reply:
a8fecf350d00d13e76f478899092d8302bfa63b3 pushed. I just realized constructing the probability array off the bat is rather pointless if all the batches consist of pandas DataFrames.